# Import necessary libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Load dataset

In [2]:
netflix_raw = pd.read_csv('../data/raw/netflix_titles.csv')

In [3]:
netflix_raw.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [4]:
netflix = netflix_raw.copy()

In [5]:
netflix.shape

(8807, 12)

In [6]:
netflix.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [7]:
netflix.isna().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [8]:
#check the percentage of NaN for all columns
nulls_df = pd.DataFrame(round(netflix.isna().sum()/len(netflix),4)*100).reset_index()
nulls_df = nulls_df.rename(columns = {'index': 'header_name',0:'percent_nulls'})
nulls_df

header_name  percent_nulls
0        show_id           0.00
1           type           0.00
2          title           0.00
3       director          29.91
4           cast           9.37
5        country           9.44
6     date_added           0.11
7   release_year           0.00
8         rating           0.05
9       duration           0.03
10     listed_in           0.00
11   description           0.00

# Data Cleaning

### Cleaning 'date_added'

In [9]:
netflix['date_added'].dtype

dtype('O')

In [10]:
#change the type of 'date_added' into datetime
netflix['date_added'] = pd.to_datetime(netflix['date_added'])
netflix['date_added']

0      2021-09-25
1      2021-09-24
2      2021-09-24
3      2021-09-24
4      2021-09-24
          ...    
8802   2019-11-20
8803   2019-07-01
8804   2019-11-01
8805   2020-01-11
8806   2019-03-02
Name: date_added, Length: 8807, dtype: datetime64[ns]

In [12]:
#since there is only 10 missing value in 'date_added', I will just drop nan
netflix = netflix[netflix['date_added'].isna()==False]

In [13]:
netflix.isna().sum()

show_id            0
type               0
title              0
director        2624
cast             825
country          830
date_added         0
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [14]:
netflix.shape

(8797, 12)

### seprate movies and tvshows in the netflix dataset 

In [15]:
movies = netflix[netflix['type']=='Movie']
movies.head()

show_id   type                             title  \
0       s1  Movie              Dick Johnson Is Dead   
6       s7  Movie  My Little Pony: A New Generation   
7       s8  Movie                           Sankofa   
9      s10  Movie                      The Starling   
12     s13  Movie                      Je Suis Karl   

                         director  \
0                 Kirsten Johnson   
6   Robert Cullen, José Luis Ucha   
7                    Haile Gerima   
9                  Theodore Melfi   
12            Christian Schwochow   

                                                 cast  \
0                                                 NaN   
6   Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7   Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9   Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                              country date_added  \
0                                       United States 2021-09-25   
6                                                 NaN 2021-09-24   
7   United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                       United States 2021-09-24   
12                            Germany, Czech Republic 2021-09-23   

    release_year rating duration  \
0           2020  PG-13   90 min   
6           2021     PG   91 min   
7           1993  TV-MA  125 min   
9           2021  PG-13  104 min   
12          2021  TV-MA  127 min   

                                           listed_in  \
0                                      Documentaries   
6                           Children & Family Movies   
7   Dramas, Independent Movies, International Movies   
9                                   Comedies, Dramas   
12                      Dramas, International Movies   

                                          description  
0   As her father nears the end of his life, filmm...  
6   Equestria's divided. But a bright-eyed hero be...  
7   On a photo shoot in Ghana, an American model s...  
9   A woman adjusting to life after a loss contend...  
12  After most of her family is murdered in a terr...

In [16]:
movies.isna().sum()

show_id           0
type              0
title             0
director        188
cast            475
country         440
date_added        0
release_year      0
rating            2
duration          3
listed_in         0
description       0
dtype: int64

In [17]:
tvshows = netflix[netflix['type']=='TV Show']
tvshows.head()

show_id     type                  title         director  \
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   
5      s6  TV Show          Midnight Mass    Mike Flanagan   

                                                cast       country date_added  \
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...  South Africa 2021-09-24   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...           NaN 2021-09-24   
3                                                NaN           NaN 2021-09-24   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...         India 2021-09-24   
5  Kate Siegel, Zach Gilford, Hamish Linklater, H...           NaN 2021-09-24   

   release_year rating   duration  \
1          2021  TV-MA  2 Seasons   
2          2021  TV-MA   1 Season   
3          2021  TV-MA   1 Season   
4          2021  TV-MA  2 Seasons   
5          2021  TV-MA   1 Season   

                                           listed_in  \
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   
5                 TV Dramas, TV Horror, TV Mysteries   

                                         description  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...  
5  The arrival of a charismatic young priest brin...

In [18]:
tvshows.isna().sum()

show_id            0
type               0
title              0
director        2436
cast             350
country          390
date_added         0
release_year       0
rating             2
duration           0
listed_in          0
description        0
dtype: int64

### cleaning 'duration' in both movies and tvshows

In [19]:
#remove strings in duration column for all movies and change the type into numerical
movies['duration'] = movies['duration'].str.replace('min','').astype(float) #due to NaN, can't change type into int
movies['duration']

/var/folders/vm/q28pb78130n5_lnd8k2t837h0000gn/T/ipykernel_80258/3397772724.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['duration'] = movies['duration'].str.replace('min','').astype(float) #due to NaN, can't change type into int


0        90.0
6        91.0
7       125.0
9       104.0
12      127.0
        ...  
8801     96.0
8802    158.0
8804     88.0
8805     88.0
8806    111.0
Name: duration, Length: 6131, dtype: float64

In [20]:
movies['duration'].dtype

dtype('float64')

In [21]:
#remove strings in duration for all tvshows
words = ['Seasons','Season']
p = '|'.join(words)

tvshows['duration'] = tvshows['duration'].str.replace(p,'').astype(int)
tvshows['duration']

/var/folders/vm/q28pb78130n5_lnd8k2t837h0000gn/T/ipykernel_80258/2161068190.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  tvshows['duration'] = tvshows['duration'].str.replace(p,'').astype(int)
/var/folders/vm/q28pb78130n5_lnd8k2t837h0000gn/T/ipykernel_80258/2161068190.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tvshows['duration'] = tvshows['duration'].str.replace(p,'').astype(int)


1       2
2       1
3       1
4       2
5       1
       ..
8795    2
8796    2
8797    3
8800    1
8803    2
Name: duration, Length: 2666, dtype: int64

In [22]:
tvshows['duration'].dtype

dtype('int64')

In [23]:
#imputing missing value with median for duration in movies
movies['duration']=movies['duration'].fillna(movies['duration'].median())
movies['duration']

/var/folders/vm/q28pb78130n5_lnd8k2t837h0000gn/T/ipykernel_80258/2819642606.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  movies['duration']=movies['duration'].fillna(movies['duration'].median())


0        90.0
6        91.0
7       125.0
9       104.0
12      127.0
        ...  
8801     96.0
8802    158.0
8804     88.0
8805     88.0
8806    111.0
Name: duration, Length: 6131, dtype: float64

In [24]:
movies['duration'].isna().sum()

0

In [25]:
tvshows['duration'].isna().sum()

0

- After imputing the missing values of duration for both movies and tvshows, I will concat it into one dataframe(netflix_new)


In [26]:
netflix_new = pd.concat([movies,tvshows], axis=0,ignore_index=True)
netflix_new.head()

show_id   type                             title  \
0      s1  Movie              Dick Johnson Is Dead   
1      s7  Movie  My Little Pony: A New Generation   
2      s8  Movie                           Sankofa   
3     s10  Movie                      The Starling   
4     s13  Movie                      Je Suis Karl   

                        director  \
0                Kirsten Johnson   
1  Robert Cullen, José Luis Ucha   
2                   Haile Gerima   
3                 Theodore Melfi   
4            Christian Schwochow   

                                                cast  \
0                                                NaN   
1  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
2  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
3  Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
4  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                             country date_added  release_year  \
0                                      United States 2021-09-25          2020   
1                                                NaN 2021-09-24          2021   
2  United States, Ghana, Burkina Faso, United Kin... 2021-09-24          1993   
3                                      United States 2021-09-24          2021   
4                            Germany, Czech Republic 2021-09-23          2021   

  rating  duration                                         listed_in  \
0  PG-13      90.0                                     Documentaries   
1     PG      91.0                          Children & Family Movies   
2  TV-MA     125.0  Dramas, Independent Movies, International Movies   
3  PG-13     104.0                                  Comedies, Dramas   
4  TV-MA     127.0                      Dramas, International Movies   

                                         description  
0  As her father nears the end of his life, filmm...  
1  Equestria's divided. But a bright-eyed hero be...  
2  On a photo shoot in Ghana, an American model s...  
3  A woman adjusting to life after a loss contend...  
4  After most of her family is murdered in a terr...

- Since there is no missing value in 'duration', i can change the type from float to int.

In [27]:
netflix_new['duration']=netflix_new['duration'].astype(int)

In [28]:
netflix_new['duration'].dtype

dtype('int64')

In [29]:
netflix_new.shape

(8797, 12)

In [30]:
netflix_new.isna().sum()

show_id            0
type               0
title              0
director        2624
cast             825
country          830
date_added         0
release_year       0
rating             4
duration           0
listed_in          0
description        0
dtype: int64

### Cleaning 'rating'

In [31]:
netflix_new['rating'].unique()

array(['PG-13', 'PG', 'TV-MA', 'TV-PG', 'TV-14', 'TV-Y', 'R', 'TV-G',
       'TV-Y7', 'G', 'NC-17', '74 min', '84 min', '66 min', 'NR', nan,
       'TV-Y7-FV', 'UR'], dtype=object)

In [32]:
netflix_new['rating'].value_counts(dropna=False)

TV-MA       3205
TV-14       2157
TV-PG        861
R            799
PG-13        490
TV-Y7        333
TV-Y         306
PG           287
TV-G         220
NR            79
G             41
TV-Y7-FV       6
NaN            4
NC-17          3
UR             3
74 min         1
84 min         1
66 min         1
Name: rating, dtype: int64

-  I will fisrt replace the NaN, then correct the wrong value name'74min','84min' and '66min'.

In [33]:
movies['rating'].value_counts(dropna=False)

TV-MA       2062
TV-14       1427
R            797
TV-PG        540
PG-13        490
PG           287
TV-Y7        139
TV-Y         131
TV-G         126
NR            75
G             41
TV-Y7-FV       5
NC-17          3
UR             3
NaN            2
74 min         1
84 min         1
66 min         1
Name: rating, dtype: int64

In [34]:
tvshows['rating'].value_counts(dropna=False)

TV-MA       1143
TV-14        730
TV-PG        321
TV-Y7        194
TV-Y         175
TV-G          94
NR             4
R              2
NaN            2
TV-Y7-FV       1
Name: rating, dtype: int64

In [35]:
netflix_new['rating'].mode()

0    TV-MA
Name: rating, dtype: object

In [36]:
movies['rating'].mode()

0    TV-MA
Name: rating, dtype: object

In [37]:
tvshows['rating'].mode()

0    TV-MA
Name: rating, dtype: object

- replace 'rating' missing value with the most frequent value since the most frequent values for movies and tvshows are same.

In [38]:
netflix_new['rating'] = netflix_new['rating'].fillna(netflix_new['rating'].mode()[0])

In [39]:
netflix_new['rating'].isna().sum()

0

In [40]:
netflix_new['rating'].value_counts(dropna=False)

TV-MA       3209
TV-14       2157
TV-PG        861
R            799
PG-13        490
TV-Y7        333
TV-Y         306
PG           287
TV-G         220
NR            79
G             41
TV-Y7-FV       6
NC-17          3
UR             3
74 min         1
84 min         1
66 min         1
Name: rating, dtype: int64

In [41]:
#correct wrong values in 'rating' with the most frequent value
netflix_new['rating'] = netflix_new['rating'].replace(['74 min','84 min','66 min'],'TV-MA')
netflix_new['rating']

0       PG-13
1          PG
2       TV-MA
3       PG-13
4       TV-MA
        ...  
8792    TV-Y7
8793    TV-PG
8794    TV-Y7
8795    TV-PG
8796    TV-Y7
Name: rating, Length: 8797, dtype: object

In [42]:
netflix_new['rating'].value_counts(dropna=False)

TV-MA       3212
TV-14       2157
TV-PG        861
R            799
PG-13        490
TV-Y7        333
TV-Y         306
PG           287
TV-G         220
NR            79
G             41
TV-Y7-FV       6
NC-17          3
UR             3
Name: rating, dtype: int64

### cleaning 'country'

In [43]:
netflix_new.isna().sum()

show_id            0
type               0
title              0
director        2624
cast             825
country          830
date_added         0
release_year       0
rating             0
duration           0
listed_in          0
description        0
dtype: int64

In [44]:
netflix_new['country'].mode()[0]

'United States'

In [45]:
#replace NaN with most frequent value
netflix_new['country']=netflix_new['country'].fillna(netflix_new['country'].mode()[0])

In [46]:
netflix_new.isna().sum()

show_id            0
type               0
title              0
director        2624
cast             825
country            0
date_added         0
release_year       0
rating             0
duration           0
listed_in          0
description        0
dtype: int64

### cleaning 'cast'

- Since cast is not important here for my designed EDA questions and I am also not gonna train models in this dataset, i will just replace it with 'Unknown'. If i drop rows, it might have influence on further data analysis accuracy since there are quite a lot NaN in 'cast'.

In [47]:
netflix_new['cast'] = netflix_new['cast'].fillna('Unknown')

In [48]:
netflix_new.isna().sum()

show_id            0
type               0
title              0
director        2624
cast               0
country            0
date_added         0
release_year       0
rating             0
duration           0
listed_in          0
description        0
dtype: int64

### cleaning 'director'

- As can be seen from the missing value summary, there are 2624 missing values in 'director'.I can't just drop the rows, instead i will replace NaN with 'Unknown'.

In [49]:
netflix_new['director'] = netflix_new['director'].fillna('Unknown')

In [50]:
netflix_new.isna().sum()

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64

In [51]:
netflix_new.head()

show_id   type                             title  \
0      s1  Movie              Dick Johnson Is Dead   
1      s7  Movie  My Little Pony: A New Generation   
2      s8  Movie                           Sankofa   
3     s10  Movie                      The Starling   
4     s13  Movie                      Je Suis Karl   

                        director  \
0                Kirsten Johnson   
1  Robert Cullen, José Luis Ucha   
2                   Haile Gerima   
3                 Theodore Melfi   
4            Christian Schwochow   

                                                cast  \
0                                            Unknown   
1  Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
2  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
3  Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
4  Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   

                                             country date_added  release_year  \
0                                      United States 2021-09-25          2020   
1                                      United States 2021-09-24          2021   
2  United States, Ghana, Burkina Faso, United Kin... 2021-09-24          1993   
3                                      United States 2021-09-24          2021   
4                            Germany, Czech Republic 2021-09-23          2021   

  rating  duration                                         listed_in  \
0  PG-13        90                                     Documentaries   
1     PG        91                          Children & Family Movies   
2  TV-MA       125  Dramas, Independent Movies, International Movies   
3  PG-13       104                                  Comedies, Dramas   
4  TV-MA       127                      Dramas, International Movies   

                                         description  
0  As her father nears the end of his life, filmm...  
1  Equestria's divided. But a bright-eyed hero be...  
2  On a photo shoot in Ghana, an American model s...  
3  A woman adjusting to life after a loss contend...  
4  After most of her family is murdered in a terr...

In [52]:
#check duplicates
netflix_new.duplicated().sum()

0

In [53]:
netflix_new.shape

(8797, 12)

In [62]:
netflix_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8797 entries, 0 to 8796
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       8797 non-null   object        
 1   type          8797 non-null   object        
 2   title         8797 non-null   object        
 3   director      8797 non-null   object        
 4   cast          8797 non-null   object        
 5   country       8797 non-null   object        
 6   date_added    8797 non-null   datetime64[ns]
 7   release_year  8797 non-null   int64         
 8   rating        8797 non-null   object        
 9   duration      8797 non-null   int64         
 10  listed_in     8797 non-null   object        
 11  description   8797 non-null   object        
dtypes: datetime64[ns](1), int64(2), object(9)
memory usage: 824.8+ KB


- Save the cleaned dataset as 'netflix_cleaned'

In [57]:
netflix_new.to_csv('../data/clean/netflix_cleaned.csv', index=False)

In [64]:
# All above is about data cleaning. The EDA part is in the other jupyter notebook.